### PR TITLE
[Merged by Bors] - fix #290 where event are not batched on linger

### LINKED ISSUE #290: Kafka sink connector does not seem to be batching
When I set `linger.ms` to 100ms, I only get one record produced every 100ms.

### PR DESCRIPTION
**Pull request by Kelerchian**
**Root cause:**
From awaiting future returned by `kafka_producer.send`, which resolves only after the record is sent, not when it is queued. Which blocks the main loop (`while let Some(Ok(record)) = stream.next().await`)

**Solution**
Use `tokio::spawn` to delegate concurrent-asynchronous polling of the `futures` to tokio's built-in runtime which is non-blocking.

**Notes/questions for posterity**
- With the blocking mechanism in the main loop gone and the usage of tokio::spawn, would we need a backpressure mechanism to avoid memory over-allocation?

**Additional notes:**
There is much change in the code because I had to flatten all the code in the main function to make sense of the lifetimes.

### CODE REVIEW

**Comment by Kelerchian** [comment]
**Manual test proof**

kafka-sink-batching

**Comment by simlay** [comment]
Thanks for the PR btw! Hopefully you learned something new and found the documentation okay.

File: rust-connectors/sinks/kafka/src/main.rs
Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs:65]: > It's a little unclear to me what happens when you don't await on any of the spawned tasks.

Without `tokios::spawn`, it would be a stagnant future - not executed by anything. Awaiting a future is essentially polling it until it yields `Poll::Ready`. Moving it using `tokio::spawn` delegates the polling to tokio's default runtime.

> Also, we don't get any errors if we don't await.

There are two futures involved here. 1.) the `async move block` and 2.) `let future = producer.send`.
Future 1 is going to be handled by the built-in tokio runtime.
Future 2 is awaited inside of Future 1, which is in our control. On line 53, the same file, we can catch `res` when it is `Err`. Error handling can be done here as follows:
We can build a batched error pool. Said error pool is processed on the side, concurrent with the main loop. On `Err() = res` the Err value can be then piped into the error pool.

File: rust-connectors/sinks/kafka/src/main.rs
Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs:46]: Oh this one is a useless artifact. My bad. I'm removing it.

Previously it was FutureProducer that requires `as`, maybe because of ambiguous type inference on rustc's part.

**Comment by Kelerchian** [comment]
> Looks like you've got some clippy lint issues though.

Fixing this right away.

> Thanks for the PR btw! Hopefully you learned something new and found the documentation okay.

I can quite follow the code easily. Though, setting up the environment to build the project and to make it work with my vscode is a bit difficult (I'm on Windows). Have you got any tips and tricks on that part?

File: rust-connectors/sinks/kafka/src/main.rs
Comment (sehz) [inline rust-connectors/sinks/kafka/src/main.rs:72]: why not combine into a single line for await?

**Review by sehz — CHANGES_REQUESTED** [review]

**Comment by simlay** [comment]
> > Thanks for the PR btw! Hopefully you learned something new and found the documentation okay.
> 
> I can quite follow the code easily. Though, setting up the environment to build the project and to make it work with my vscode is a bit difficult (I'm on Windows). Have you got any tips and tricks on that part?

Building the project is definitely more difficult than it should be. Given that the connectors are meant to run in a linux docker container, which means it's cross compiling. No one's actually tried to build/run this environment with windows so kudos to you for making it work.

File: rust-connectors/sinks/kafka/src/main.rs
Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs:72]: Leftover code that I missed. I tried passing the future to the tokio::spawn but there was the Arc problem. Will make it single line

**Comment by Kelerchian** [comment]
> > > Thanks for the PR btw! Hopefully you learned something new and found the documentation okay.
> > 
> > 
> > I can quite follow the code easily. Though, setting up the environment to build the project and to make it work with my vscode is a bit difficult (I'm on Windows). Have you got any tips and tricks on that part?
> 
> Building the project is definitely more difficult than it should be. Given that the connectors are meant to run in a linux docker container, which means it's cross compiling. No one's actually tried to build/run this environment with windows so kudos to you for making it work.

Clarifying. It was windows + WSL :D, so it was pretty much linux.
Have you considered Docker-based builder so that it just runs everywhere without much setup?

File: rust-connectors/sinks/kafka/src/main.rs
Comment (simlay) [inline rust-connectors/sinks/kafka/src/main.rs:73]: I think we should do something like this (even though it's a little out of the scope of this PR):
```suggestion
            if let Err(e) = producer.send(kafka_record, Duration::from_secs(0)).await {
                error!("Kafka produce {:?}", e)
            }
```
I shouldn't have left the `debug!` line in when I originally build the connector. The debug output is pretty useless.

File: rust-connectors/sinks/kafka/src/main.rs
Comment (simlay) [inline rust-connectors/sinks/kafka/src/main.rs:3]:
```suggestion
use fluvio_future::tracing::{debug, info, error};
```

This for https://github.com/infinyon/fluvio-connectors/pull/306/files#r927820467

File: rust-connectors/sinks/kafka/src/main.rs
Comment (sehz) [reply rust-connectors/sinks/kafka/src/main.rs:64]: There is a fundamental problem here.  Since this spawns new future tasks, it could cause an unlimited number of tasks that could cause OOM or other issues.  Not only that, if Kafka goes down or returns some other error, this would just create an endless number of tasks that just spit out errors.

**Review by sehz — CHANGES_REQUESTED** [review]

File: rust-connectors/sinks/kafka/src/main.rs
Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs:64]: I agree this requires a backpressure mechanism. I'll search for it then. My intuition is to add a limit to the max number of task in a tokio pool. I can't find the API to get that count, so I think I would need to implement that with a pool of JoinHandle and select! inside the main loop to context switch between 1.) the piping of events to kafka and 2.) "garbage-collecting" the finished JoinHandles in the pool

Btw, CMIIW, I assume tokio's built-in would automatically drop the resolved futures regardless the JoinHandle is awaited or not. Would you know if that assumption is correct or not? I'll ask the forum next about it.

Comment (sehz) [reply rust-connectors/sinks/kafka/src/main.rs:64]: I don't think it's a good idea to rely on Tokio pool since we might move to another engine in the future. Instead, it would be better to build application-specific code to deal with this.

Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs:64]: Ok, what about a pool of pairs of JoinHandle and a sum of approximate memory usage of the related future

The additional parameter would be the max memory usage (with caveat that the calculation is an approximate and we describe how we are calculating it and the limitation and pitfall of the parameter)

Comment (sehz) [reply rust-connectors/sinks/kafka/src/main.rs:64]: How about using a channel?

Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs:64]: > How about using a channel?

Sorry, I don't follow. Could you elaborate more on using channel?

Btw, the solution I'm thinking that utilize the approximate memory usage looks like this https://github.com/Kelerchian/fluvio-connectors/tree/290-kafka-sink-batch-with-backpressure

Disclaimer, that branch is not finished yet

Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs:64]: (I commented several days ago but it was "pending" and did not realize what it meant, probably you couldn't read that one)

**Comment by sehz** [comment]
@galibey can you review this as well?

**Comment by galibey** [comment]
> **Root cause:** From awaiting future returned by `kafka_producer.send`, which resolves only after the record is sent, not when it is queued. Which blocks the main loop (`while let Some(Ok(record)) = stream.next().await`)

There is a method `FutureProducer::send_result` which resolves when the record is queued and returns `DeliveryFuture` that you may resolve if you want to wait for the actual saved result. We can use it inside loop (`while let Some(Ok(record)) = stream.next().await`). If we want to periodically check if records are actually sent and acked we can periodically resolve the latest `DeliveryFuture` (essentially implement our own linger). This would be a more straightforward approach than having to worry about memory and back pressure mechanisms.

**Comment by Kelerchian** [comment]
> > **Root cause:** From awaiting future returned by `kafka_producer.send`, which resolves only after the record is sent, not when it is queued. Which blocks the main loop (`while let Some(Ok(record)) = stream.next().await`)
> 
> There is a method `FutureProducer::send_result` which resolves when the record is queued and returns `DeliveryFuture` that you may resolve if you want to wait for the actual saved result. We can use it inside loop (`while let Some(Ok(record)) = stream.next().await`). If we want to periodically check if records are actually sent and acked we can periodically resolve the latest `DeliveryFuture` (essentially implement our own linger). This would be a more straightforward approach than having to worry about memory and back pressure mechanisms.

Oh, utilizing rdkafka's buffering mechanism. That's neat, thanks.

 I'm not too well-versed with tokio, but I assume it will need to "park" for a while on "queue-full" error. Is it good to use `sleep(Duration::from_millis(0)).await;` for it?

Also, for sinks, what's the policy for error handling? Do we retry unsent messages or do we ignore those that is caused by some classes of errors?

**Comment by galibey** [comment]
> I'm not too well-versed with tokio, but I assume it will need to "park" for a while on "queue-full" error. Is it good to use `sleep(Duration::from_millis(0)).await;` for it?
If you are talking about `FutureProducer::send_result`, it delegates queueing to the native `librdkafka` library. If it blocks in case of a full queue, it's ok, I think we should not try to optimize it away on our app level.


> Also, for sinks, what's the policy for error handling? Do we retry unsent messages or do we ignore those that is caused by some classes of errors?
There is no policy per se. Retry mechanism is a big topic on its own and should be handled in a separate issue.

**Comment by Kelerchian** [comment]
> If you are talking about `FutureProducer::send_result`, it delegates queueing to the native `librdkafka` library. If it blocks in case of a full queue, it's ok, I think we should not try to optimize it away on our app level.

Sorry I wasn't clear about it. I was talking about tokio's internal mechanism for switching between its green threads. In a single threaded context, yielding in one green thread is required to notify tokio's scheduler that the opportunity to context-switch is now, and I used "tokio::time::sleep" for that.

Otherwise, tokio's scheduler won't be able to switch to future_producer's poller. That will deadlock the application if the queue is full and the the application is stuck inside the inside-most while loop.

After reading [this](https://users.rust-lang.org/t/tokio-context-switching-for-spawn-vs-spawn-blocking/63571/5) though, I'll change the implementation to use `tokio::spawn` instead for the sake of clarity.

multithread-vs-singlethread

**Comment by Kelerchian** [comment]
> There is no policy per se. Retry mechanism is a big topic on its own and should be handled in a separate issue.

Alright then. I'll stick to `error!` ing the `Result::Err`

**Comment by galibey** [comment]
> Otherwise, tokio's scheduler won't be able to switch to future_producer's poller. That will deadlock the application if the queue is full and the the application is stuck inside the inside-most while loop.

Putting `FutureProducer::send_result` inside while loop is dangerous, and if you do, `sleep` should be longer, otherwise, if something went wrong the loop will eat CPU a lot.
I would recommend, just fail-fast if there is an error until we get more robust retry and error handling logic.

File: rust-connectors/sinks/kafka/src/main.rs
Comment (galibey) [inline rust-connectors/sinks/kafka/src/main.rs:66]: I think we should stop processing here (ideally with printing the last read offset) instead of having gaps in the sent records sequence.

File: rust-connectors/sinks/kafka/src/main.rs
Comment (galibey) [inline rust-connectors/sinks/kafka/src/main.rs:13]: This comment block seems redundant to me. 
If we want to comment on something, let's just do it without graphical wrappings.

File: rust-connectors/sinks/kafka/src/main.rs
Comment (galibey) [inline rust-connectors/sinks/kafka/src/main.rs:36]: This comment block seems also redundant to me.

File: rust-connectors/sinks/kafka/src/main.rs
Comment (galibey) [inline rust-connectors/sinks/kafka/src/main.rs:45]: This comment block seems also redundant to me.

File: rust-connectors/sinks/kafka/src/main.rs
Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs:66]: Ok. Changing it. Adding "last_record_read" in the form of json line so that it is machine readable.
Also, what happens to connectors that exit?

File: rust-connectors/sinks/kafka/src/main.rs
Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs]: removing

File: rust-connectors/sinks/kafka/src/main.rs
Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs]: removing

File: rust-connectors/sinks/kafka/src/main.rs
Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs]: removing

File: rust-connectors/sinks/kafka/src/main.rs
Comment (galibey) [inline rust-connectors/sinks/kafka/src/main.rs:62]: We construct a `Value` just to serialize it again in the next line, this is not efficient. We could just format a string immediately. For example, `error!("KafkaError {:?}, offset: {}, partition: {}", error, &record.offset, &record.partition);`

Why do we need it in JSON format, btw? and why `println`?

Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs:62]: JSON line is machine-readable. It is useful for automation in the future. Is it unnecessary?

Comment (galibey) [reply rust-connectors/sinks/kafka/src/main.rs:62]: For automation purposes, it's likely should be done for the entire program. Usually, it is configured at the logger level once, and no need to deal with it in every log printing.

Comment (Kelerchian) [reply rust-connectors/sinks/kafka/src/main.rs:62]: What I mean by automation is for other components that may consume the `stdout/stderr` (e.g when piped into something like kibana, these lines of output will be a bit easier to digest rather than an unstructured string, or scheduler that may have to manage the runtime of the connectors). 
> `println!`
My bad, it should have used something that puts to `stderr`

Either way, those are things out of the scope of this issue. I'll revert it back to using `error!`

Comment (galibey) [reply rust-connectors/sinks/kafka/src/main.rs]: Yes, but for Kibana would be useful to get in a structured way not only the message itself but also a timestamp, module, log level, etc. 
Instead of: `2022-08-03T08:43:38Z INFO main some message` 
parse
`{
   "timestamp": "2022-08-03T08:43:38Z",
   "level": "INFO",
   "module": "main",
   "body": "main some message"
} 
`
right?
And that better to be done in one place. For example, using custom formatting like https://docs.rs/env_logger/0.9.0/env_logger/#using-a-custom-format, if there are no out-of-box solutions for JSON

**Comment by galibey** [comment]
Looks good to me. Please wait for approval from @simlay , as he is the main maintainer of Kafka Sink connector.

**Review by simlay — APPROVED** [review]
Yeah, this looks good. I think we've resolved all the unanswered issues.

**Pull request closed without merging**
